language/move-core/types/src/effects.rs

```diff
@@ -7,60 +7,114 @@ use crate::{
+    /// Deletes some data that currently exists.
+    Delete,
 }
 
-fn publish_checked<K, V, F>(map: &mut BTreeMap<K, Option<V>>, k: K, v: V, make_err: F) -> Result<()>
-where
-    K: Ord,
-    F: FnOnce() -> Error,
-{
-    match map.entry(k) {
-        btree_map::Entry::Occupied(entry) => {
-            let r = entry.into_mut();
-            match r {
-                Some(_) => return Err(make_err()),
-                None => *r = Some(v),
-            }
+impl<T> Op<T> {
+    pub fn as_ref(&self) -> Op<&T> {
+        use Op::*;
+
+        match self {
+            New(data) => New(data),
+            Modify(data) => Modify(data),
+            Delete => Delete,
         }
-        btree_map::Entry::Vacant(entry) => {
-            entry.insert(Some(v));
+    }
+
+    pub fn map<F, U>(self, f: F) -> Op<U>
+    where
+        F: FnOnce(T) -> U,
+    {
+        use Op::*;
+
+        match self {
+            New(data) => New(f(data)),
+            Modify(data) => Modify(f(data)),
+            Delete => Delete,
         }
     }
-    Ok(())
+
+    pub fn ok(self) -> Option<T> {
+        use Op::*;
+
+        match self {
+            New(data) | Modify(data) => Some(data),
+            Delete => None,
+        }
+    }
+}
+
+/// A collection of resource and module operations on a Move account.
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct AccountChangeSet {
+    modules: BTreeMap<Identifier, Op<Vec<u8>>>,
+    resources: BTreeMap<StructTag, Op<Vec<u8>>>,
 }
 
-fn unpublish_checked<K, V, F>(map: &mut BTreeMap<K, Option<V>>, k: K, make_err: F) -> Result<()>
+/// This implements an algorithm to squash two change sets together by merging pairs of operations
+/// on the same item together. This is similar to squashing two commits in a version control system.
+///
+/// It should be noted that all operation types have some implied pre and post conditions:
+///   - New
+///     - before: data doesn't exist
+///     - after: data exists (new)
+///   - Modify
+///     - before: data exists
+///     - after: data exists (modified)
+///   - Delete
+///     - before: data exists
+///     - after: data does not exist (deleted)
+///
+/// It is possible to have a pair of operations resulting in conflicting states, in which case the
+/// squash will fail.
+fn squash<K, V>(map: &mut BTreeMap<K, Op<V>>, other: BTreeMap<K, Op<V>>) -> Result<()>
 where
     K: Ord,
-    F: FnOnce() -> Error,
 {
-    match map.entry(k) {
-        btree_map::Entry::Occupied(entry) => {
-            let r = entry.into_mut();
-            match r {
-                Some(_) => *r = None,
-                None => return Err(make_err()),
+    use btree_map::Entry::*;
+    use Op::*;
+
+    for (key, op) in other.into_iter() {
+        match map.entry(key) {
+            Occupied(mut entry) => {
+                let r = entry.get_mut();
+                match (r.as_ref(), op) {
+                    (Modify(_) | New(_), New(_)) | (Delete, Delete | Modify(_)) => {
```

Comment:
I'm wondering if `(New(d1), New(d2)` is really a conflict. You define this such that the right side overrides the left. Why would this then not be resolved as `New(d2)`? Same for other combinations.

Reply:
It is indeed a conflict. The current semantics of squash is that the resulting `ChangeSet` should be equivalent to applying the two `ChangeSet`s in serial. This means after the first `New`, the slot will be occupied and the second `New` shall fail since it implies no old data exists.

It might be easier to reason about this in terms of version control. You have 3 commits: base, c1 and c2 where c1 and c2 are children of base. Squashing c1 and c2 is actually a two-step process: you first rebase c2 on top of c1 and then squash the contents. In this case, c1 and c2 both modifies the same content so the rebase will fail due to a merge conflict.